packages/core/src/tracing/instrumentation/messaging/rdkafka.js

```diff
@@ -0,0 +1,412 @@
+   * the message was delivered successfully.
+   */
+  const deliveryCb =
+    ctx._cb_configs &&
+    ctx._cb_configs.event &&
+    ctx._cb_configs.event.delivery_cb &&
+    typeof ctx._cb_configs.event.delivery_cb === 'function';
+
+  return cls.ns.runAndReturn(() => {
+    const span = cls.startSpan('kafka', constants.EXIT);
+    const topic = originalArgs[0];
+
+    span.stack = tracingUtil.getStackTrace(instrumentedProduce, 1);
+    span.data.kafka = {
+      service: topic,
+      access: 'send'
+    };
+
+    const headers = addTraceContextHeader(originalArgs[6], span);
+    originalArgs[6] = headers;
+
+    if (deliveryCb) {
+      ctx.once('delivery-report', err => {
+        span.d = Date.now() - span.ts;
+        if (err) {
+          span.ec = 1;
+          span.data.kafka.error = err.message;
+        }
+        span.transmit();
+      });
+    }
+
+    try {
+      const result = originalProduce.apply(ctx, originalArgs);
+
+      if (!deliveryCb) {
+        span.d = Date.now() - span.ts;
+        span.transmit();
+      }
+
+      return result;
+    } catch (error) {
+      // This error is properly captured for both Stream and standard API
+      span.ec = 1;
+      span.data.kafka.error = error.message;
+
+      if (!deliveryCb) {
+        span.d = Date.now() - span.ts;
+        span.transmit();
+      }
+
+      throw error;
+    }
+  });
+}
+
+function removeInstanaHeadersFromMessage(messageData) {
+  if (messageData.headers && messageData.headers.length) {
+    const instanaHeaders = [
+      constants.kafkaTraceLevelHeaderNameString,
+      constants.kafkaTraceLevelHeaderNameBinary,
+      constants.kafkaTraceIdHeaderNameString,
+      constants.kafkaSpanIdHeaderNameString,
+      constants.kafkaTraceContextHeaderNameBinary
+    ].map(header => header.toLowerCase());
+
+    for (let i = messageData.headers.length - 1; i >= 0; i--) {
+      const headerObject = messageData.headers[i];
+      // There should be only one. That's how the API works. This can be tested, for instance, by sending a message
+      // via kafkajs to be recieved by rdkafka consumer. Each header will be a single object as an item in the
+      // messageData.headers array
+      const headerKey = Object.keys(headerObject)[0];
+
+      if (instanaHeaders.indexOf(headerKey.toLocaleLowerCase()) > -1) {
+        messageData.headers.splice(i, 1);
+      }
+    }
+  }
+}
+
+function instrumentedEmit(ctx, originalEmit, originalArgs) {
+  let [event, eventData] = originalArgs;
+
+  /**
+   * If the stream batch is enabled, eventData will be an array. So in order to guarantee that array and non array cases
+   * are covered, we make sure that the data is always an array, and loop through them at all times.
+   */
+  if (!Array.isArray(eventData)) {
+    eventData = [eventData];
+  }
+
+  if (isActive) {
+    eventData.forEach(messageData => {
+      const parentSpan = cls.getCurrentSpan();
+
+      if (!parentSpan) {
```

Comment:
I was wondering why we do not call the original emit function for both cases:

- isActive is false (https://github.com/instana/nodejs/pull/486/files#diff-24bbb2057190b9840c5c35b03ff3849adf0501722464d3ba476378bf0dfd8fd4R211)
- parentSpan is not null (https://github.com/instana/nodejs/pull/486/files#diff-24bbb2057190b9840c5c35b03ff3849adf0501722464d3ba476378bf0dfd8fd4R215)

Reply:
Nice catch! I don't know why this did not cause errors before, as it should.
But they were missing for sure.
Thanks!

Reply:
nitpick/opinion/matter of taste: I would marginally favor an early return for the cases that Kate mentioned. 

That is, something like

```
if (!isActive) {
   return originalEmit.apply(ctx, originalArgs);
}
if (parentSpan) {
   return originalEmit.apply(ctx, originalArgs);
}
if (cls.tracingSuppressed()) {
       const headers = addTraceLevelSuppression(messageData.headers);
        messageData.headers = headers;
        return originalEmit.apply(ctx, originalArgs);
}

... rest of the instrumentation logic
```

However, really just my opinion/preference, not a blocker or anything.

Reply:
I will take this over if it's okay for you @willianpc